pkg/columnstore/granule.go

```diff
@@ -62,62 +63,67 @@ func NewGranule(granulesCreated prometheus.Counter, schema *Schema, parts ...*Pa
 }
 
 // AddPart returns the new cardinality of the Granule
-func (g *Granule) AddPart(p *Part) int {
-	g.Lock()
-	defer g.Unlock()
+func (g *Granule) AddPart(p *Part) uint64 {
 
-	g.parts = append(g.parts, p)
-	g.card += p.Cardinality
+	node := g.parts.Prepend(p)
+	newcard := atomic.AddUint64(&g.card, uint64(p.Cardinality))
 	it := p.Iterator()
 
 	if it.Next() {
-		r := Row{Values: it.Values()}
-		if g.schema.RowLessThan(r.Values, g.least.Values) {
-			g.least = r
+		for {
+			r := &Row{Values: it.Values()}
+			least := atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&g.least)))
```

Comment:
Nice, lots of unsafe. :smile: 
Would be good to document why we need the magic here.